Source/WebCore/rendering/RenderBlockFlow.cpp

```diff
@@ -1193,12 +1214,33 @@ void RenderBlockFlow::determineLogicalLeftPositionForChild(RenderBox& child, App
     // If the child is being centred then the margin calculated to do that has factored in any offset required to
     // avoid floats, so use it if necessary.
 
-    if (style().textAlign() == TextAlignMode::WebKitCenter || child.style().marginStartUsing(&style()).isAuto())
+    if (style().textAlign() == TextAlignMode::WebKitCenter || child.style().marginStartUsing(&style()).isAuto()) {
         newPosition = std::max(newPosition, positionToAvoidFloats + childMarginStart);
-    else if (positionToAvoidFloats > initialStartPosition)
+        setLogicalLeftForChild(child, style().isLeftToRightDirection() ? newPosition : totalAvailableLogicalWidth - newPosition - logicalWidthForChild(child), applyDelta);
+        return;
+    }
+
+    if (positionToAvoidFloats > initialStartPosition)
         newPosition = std::max(newPosition, positionToAvoidFloats);
+    newPosition = computeSelfAlignedLogicalLeftPositionForChild(newPosition);
+    setLogicalLeftForChild(child, newPosition, applyDelta);
+}
 
-    setLogicalLeftForChild(child, style().isLeftToRightDirection() ? newPosition : totalAvailableLogicalWidth - newPosition - logicalWidthForChild(child), applyDelta);
+void RenderBlockFlow::applyStretchAlignmentToChildIfNeeded(RenderBox& child)
+{
+    auto hasAutoLogicalWidth = isHorizontalWritingMode() ? child.style().width().isAuto() : child.style().height().isAuto();
+    auto hasAutoMarginsInInlineDirection = [&] {
+        if (isHorizontalWritingMode())
+            return child.style().marginLeft().isAuto() || child.style().marginRight().isAuto();
+        return child.style().marginTop().isAuto() || child.style().marginBottom().isAuto();
+    }();
+    auto childSelfAlignmentStyle = child.style().resolvedJustifySelf(&style(), ItemPosition::Normal);
+    auto allowedToStretchChild = childSelfAlignmentStyle.position() == ItemPosition::Stretch && hasAutoLogicalWidth && !hasAutoMarginsInInlineDirection;
```

Comment:
Run the `stretch` check first and short-circuit the rest of the function if it's false. No need to run any of this logic if it's e.g. normal (which is the default, and therefore the most likely case).